### DISCUSSION TITLE
Passing an Authorization Token to the tool & avoiding the LLM.

### DISCUSSION THREAD

**Discussion by fatimatayeb** [langchain-ai/langserve#534] [Q&A]
I am using Langserve Conversation Agent with GPT models to map user questions to private API's and do function calling. 

I want to be able to **pass the authorization token directly to the tool** instead of it going through the LLM. 
Could you point me in to the right direction to do this and what is the best practice?

**Comment by eyurtsev** [comment]
The general strategy would be to sub-class form StructuredTools, implement some configurable tools for tokens, and use configurable runnables. However, because agent executor does not propagate configuration information currently, this strategy will not work. (Issue needs to be fixed in langchain)

The way to do this right now is to instantiate an appropriate agent at run time with auth bound to the tools. Instantiation of most of these objects at run time is very fast (should be far less than ~1ms), so performance is not a concern here.

Here are two reference examples that show how an agent can be instantiated at run time:

- https://github.com/langchain-ai/langserve/blob/main/examples/agent_custom_streaming/server.py
- https://github.com/langchain-ai/langserve/tree/main/examples/configurable_agent_executor

----

If the token is supplied directly by the user, the examples are sufficient as they are.

If the token comes from database look up that relies on the user identity, take a look at the auth examples here: https://python.langchain.com/docs/langserve#examples 

If you're comfortable with FastAPI / python, you can use `APIHandler` directly instead of `add_routes` for low level access and define the endpoints manually.

----

Here's documentation for configurable runnables:

https://python.langchain.com/docs/expression_language/how_to/configue

Configurable runnables examples: 

https://github.com/langchain-ai/langserve/blob/main/examples/configurable_chain/server.py

**Comment by fatimatayeb** [comment]
Lets say the auth is passed to the endpoint  like this: 

``` python
class Input(BaseModel):
    question: str
    token: str
```

I want to pass the question to the LLM but once the corresponding function is mapped, I want to intercept the function call and pass the token before it is executed. 

It is unclear how I can pass the token directly to the mapped function from the examples shared. 

I attempted to write a custom parser and pass the token to it but I am not able to extract the token from the input request as it gets overwritten by the previous step ( I am not an expert with LCEL do point out if there is a better way to do this, I tried using RunnablePassthrough with RunnableParallel still didn't get the result I wanted)

```python 
agent = (
 {
       "input": lambda x: x["question"],
        "agent_scratchpad": lambda x: format_to_openai_functions(x["intermediate_steps"]),
     }
     | prompt
     | llm_with_tools
     | CustomOpenAIFunctionsAgentOutputParser(config_param={"token" :lambda x: x["token"]}) # Pass the "token" key
 )
 )
```

Would appreciate a nudge to the right direction @eyurtsev

**Reply by fatimatayeb** [reply]
What is the best way to pass a variable coming from the user straight to the parser (modifying the the function arguments before they are invoked)? 

The method above does not work because the "token" value  is not  in the output of the previous step, how do I call the "token" value in the parser step? @eyurtsev

**Comment by casperakos** [comment]
Any updates on this? Have the same issue, @eyurtsev would appreciate any help!

**Comment by bukacdan** [comment]
I'm currently struggling with this too, I'll provide my solution if I find one. I also think that this is a very common pattern, when you, in your tools, want to access some data coming from the request without passing them through the chain. So, I believe it should have an elegant solution within the framework, along with a clear, end-to-end example explaining the principles.

**Comment by mathlover777** [comment]
I am also facing a similar usecase and would appreciate help. the bot suggested multiple approaches but did not work https://github.com/langchain-ai/langchain/discussions/22529

**Comment by rvndbalaji** [comment]
I was able to solve this problem in the following way for a Service API that requires a bearer token

**Create the required tool by calling the wrapper func**


```python
api_service = wrap_api_service({'secret' : '358123'})

tools = [api_service]

#Pass the `tools` to an agent executor or an LLM directly 

```


**Definition of the wrapper func and the tool**

```python
from typing import Callable, Any, Dict
from functools import wraps

def wrap_api_service(wrapper_config : Dict[str, Any]):
    @tool
    def api_service(request_body) -> str:

        #Wrapper config will contain the secret
        response = http_utils.post(url, request_body, wrapper.config['secret'])
        return str(response)

    return api_service


```


FYI - @fatimatayeb , @casperakos  @bukacdan @eyurtsev 

Credits - `Claude 3.5 `

**Comment by ebrahim-2** [comment]
If anyone is looking for a way to dynamically inject a JWT token into the tools during runtime, Python's `contextvars` is a handy solution. It allows you to manage context-specific data at runtime.

## Example:
#### Create a Context Manager for Token Management
```python
from contextvars import ContextVar
from typing import Optional


jwt_token: ContextVar[Optional[str]] = ContextVar("jwt_token", default=None)

class JWTTokenManager:
    def __init__(self, token: str):
        self.token = token
        self.token_context = None

    def __enter__(self):
        # Set the token in the context variable
        self.token_context = jwt_token.set(self.token)

    def __exit__(self, exc_type, exc_val, exc_tb):
        # Reset the context variable after use
        jwt_token.reset(self.token_context)
```

#### Create a Decorator to Inject the Token Into LangChain Tools
```python
from functools import wraps
from context import jwt_token

def inject_jwt_token(func):
    @wraps(func)
    async def wrapper(*args, **kwargs):
        # Retrieve the token from the context variable
        token = jwt_token.get()
        if token:
            # Inject the token into the headers
            kwargs["headers"] = kwargs.get("headers", {})
            kwargs["headers"]["Authorization"] = f"Bearer {token}"
        return await func(*args, **kwargs)
    return wrapper
```

#### Use the Decorator in a LangChain Tools
```python
from langchain_core.tools import tool
import httpx
from decorators import inject_jwt_token

@tool()
@inject_jwt_token
async def make_request(endpoint: str, headers: dict = None) -> dict:

    async with httpx.AsyncClient() as client:
        response = await client.get(endpoint, headers=headers)
        response.raise_for_status()
        return response.json()
```

#### Usage
The `JWTTokenManager` sets the token in the context. When `make_request` is called, the headers with the token is automatically added to the function.
```python
from token_manager import JWTTokenManager
from tools import make_request

async def main():
    # Set the token using the context manager
    with JWTTokenManager("your_jwt_token_here"):
        # Use the tool to make an API request
        result = await make_request("https://api.example.com/data")
        print(result)
```

**Comment by jamiecrobb** [comment]
I also would find this really useful - any updates? 

Trying to pass database user_id through AgentExecutor to a tool so that it will only edit the correct database rows.

**Reply by rvndbalaji** [reply]
I was able to solve this problem in the following way for a Service API that requires a bearer token

Create the required tool by calling the wrapper func
```

api_service = wrap_api_service({'secret' : '358123'})

tools = [api_service]

```
#Pass the `tools` to an agent executor or an LLM directly 
Definition of the wrapper func and the tool
```

from typing import Callable, Any, Dict
from functools import wraps

def wrap_api_service(wrapper_config : Dict[str, Any]):
    @tool
    def api_service(request_body) -> str:

        #Wrapper config will contain the secret
        response = http_utils.post(url, request_body, wrapper.config['secret'])
        return str(response)

    return api_service
```

**Reply by Dhupmaps** [reply]
This is a fantastic question and a perfect candidate for earning your Galaxy Brain badge (answering a "Best Practice" architectural question).

The best practice here is to inject the token at runtime via RunnableConfig, rather than letting the LLM "see" or generate the token. This keeps the sensitive key out of the prompt (security) and out of the LLM's context window (efficiency).

Here is the step-by-step implementation for LangServe:

Step 1: Define the Tool to Accept RunnableConfig
In LangChain, if you type hint a tool argument with RunnableConfig, the LLM will ignore it (it won't try to generate it), but the runtime will automatically inject the configuration object when the tool executes.

```
from langchain_core.tools import tool
from langchain_core.runnables import RunnableConfig

@tool
def private_api_tool(query: str, config: RunnableConfig) -> dict:
    """Call the private API with the user's secure token."""
    
    # 1. Retrieve the token safely from the runtime configuration
    # We will populate this 'auth_token' in Step 2
    user_token = config.get("configurable", {}).get("auth_token")
    
    if not user_token:
        return {"error": "Authentication token missing"}

    # 2. Make your API call using the token
    # response = requests.get(url, headers={"Authorization": f"Bearer {user_token}"})
    
    return {"status": "success", "data": f"Authorized access for query: {query}"}
    
```
    Step 2: Extract Token in LangServe (per_req_config_modifier)
When setting up your LangServe route, you need a way to grab the Bearer token from the incoming HTTP request and shove it into that configurable dictionary so the tool can find it.

LangServe provides a hook called per_req_config_modifier exactly for this purpose.

```
from typing import Dict, Any
from fastapi import Request

def inject_token(config: Dict[str, Any], request: Request) -> Dict[str, Any]:
    """
    Extracts the Authorization header from the incoming LangServe request
    and injects it into the configurable fields.
    """
    # Get the token from headers (e.g., "Authorization: Bearer xyz")
    auth_header = request.headers.get("Authorization")
    
    if auth_header:
        # Simple parsing to remove "Bearer " prefix if present
        token = auth_header.replace("Bearer ", "").strip()
        
        # Inject into config
        config.setdefault("configurable", {})
        config["configurable"]["auth_token"] = token
        
    return config
```
    
    Step 3: Connect it in server.py
Finally, wire it all together in your FastAPI app.

```
from langserve import add_routes
from fastapi import FastAPI
# Import your agent_executor (which uses the tool above)

app = FastAPI()

add_routes(
    app,
    agent_executor, # Your agent that uses private_api_tool
    path="/my-agent",
    # This magic hook connects the HTTP Request -> RunnableConfig
    per_req_config_modifier=inject_token 
)
```

Why this is the Best Practice:
Security: The token never enters the prompt. Even if the LLM hallucinates, it cannot leak the credential because it never saw it.

Audit Logs: Your LLM input/output logs (in LangSmith) will show the tool call happening, but the token remains hidden in the separate config layer.

Statelessness: This works perfectly for multi-user environments where every request has a different token.
